test/python/circuit/test_hamiltonian_gate.py

```diff
@@ -0,0 +1,201 @@
+
+# pylint: disable=arguments-differ,method-hidden
+
+"""Quick program to test the qi tools modules."""
+
+import json
+import numpy as np
+from numpy.testing import assert_allclose
+
+
+import qiskit
+from qiskit.extensions.hamiltonian_gate import HamiltonianGate, UnitaryGate
+from qiskit.extensions.exceptions import ExtensionError
+from qiskit.test import QiskitTestCase
+from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.quantum_info import Operator
+from qiskit.transpiler import PassManager
+from qiskit.compiler import transpile
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.transpiler.passes import CXCancellation
+
+
+class TestHamiltonianGate(QiskitTestCase):
+    """Tests for the HamiltonianGate class."""
+
+    def test_set_matrix(self):
+        """Test instantiation"""
+        hamiltonian = HamiltonianGate([[0, 1], [1, 0]], 1)
+        self.assertEqual(hamiltonian.num_qubits, 1)
+
+    def test_set_matrix_raises(self):
+        """test non-unitary"""
+        with self.assertRaises(ExtensionError):
+            HamiltonianGate([[1, 0], [1, 1]], 1)
+
+    def test_conjugate(self):
+        """test conjugate"""
+        ham = HamiltonianGate([[0, 1j], [-1j, 2]], np.pi / 4)
+        np.testing.assert_array_almost_equal(ham.conjugate().to_matrix(),
+                                             np.conj(ham.to_matrix()))
+
+    def test_transpose(self):
+        """test transpose"""
+        ham = HamiltonianGate([[15, 1j], [-1j, -2]], np.pi / 7)
+        np.testing.assert_array_almost_equal(ham.transpose().to_matrix(),
+                                             np.transpose(ham.to_matrix()))
+
+    def test_adjoint(self):
+        """test adjoint operation"""
+        ham = HamiltonianGate([[3, 4j], [-4j, -.2]], np.pi * .143)
+        np.testing.assert_array_almost_equal(ham.adjoint().to_matrix(),
+                                             np.transpose(np.conj(ham.to_matrix())))
+
+
+class TestHamiltonianCircuit(QiskitTestCase):
+    """Hamiltonian gate circuit tests."""
+
+    def test_1q_hamiltonian(self):
+        """test 1 qubit hamiltonian"""
+        qr = QuantumRegister(1)
+        cr = ClassicalRegister(1)
+        qc = QuantumCircuit(qr, cr)
+        matrix = np.zeros((2, 2))
+        qc.x(qr[0])
+        theta = Parameter('theta')
+        qc.append(HamiltonianGate(matrix, theta), [qr[0]])
+        qc = qc.bind_parameters({theta: 1})
+
+        # test of qasm output
+        self.log.info(qc.qasm())
+        ref_qasm = "OPENQASM 2.0;\n" \
+                   "include \"qelib1.inc\";\n" \
+                   "qreg q0[1];\n" \
+                   "creg c0[1];\n" \
+                   "x q0[0];\n" \
+                   "hamiltonian([[0.+0.j 0.+0.j]\n" \
+                   " [0.+0.j 0.+0.j]],1) q0[0];\n"
+        self.assertEqual(qc.qasm(), ref_qasm)
+
+        # test of text drawer
+        self.log.info(qc)
```

Comment:
You can assert on the drawn circuit including a HamiltonianGate (see examples in `test/python/visualization/test_circuit_text_drawer.py`).

Reply:
Good catch - this prints pretty wonky (below) because it prints with the Hermitian matrix. Do you know how to get it to print with just the name (and maybe parameter name?) like Unitary does?
```
      ┌───┐┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
q1_0: ┤ X ├┤0                                                                                                                                                            ├
      └───┘│  Hamiltonian([[0.+0.j 0.+0.j 0.+0.j 0.-1.j]
 [0.+0.j 0.+0.j 0.+1.j 0.+0.j]
 [0.+0.j 0.-1.j 0.+0.j 0.+0.j]
 [0.+1.j 0.+0.j 0.+0.j 0.+0.j]],1.57079632679490) │
q1_1: ─────┤1                                                                                                                                                            ├
           └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
c1_0: ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
                                                                                                                                                                          
c1_1: ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
```

Reply:
By adding a label (`qc.append(HamiltonianGate(matrix, theta, label='Hamiltonian'), [qr[0]])` you will get the following result:
```
      ┌───┐┌─────────────┐
q0_0: ┤ X ├┤ Hamiltonian ├
      └───┘└─────────────┘
c0_0: ════════════════════
```
Not sure if that helps.

Reply:
Ok - so should we say this is WAI? Or should I add a label in the constructor if none is provided? Similar to: `self._qasm_name = self.label if self.label else "unitary" + str(id(self))` in the Unitary qasm generator?

Reply:
Or only add the label if none is given in the `qc.hamiltonian()` monkeypatch?

Reply:
Or just change the default label in the signature from None to "Hamiltonian"?

Reply:
I think this is WAI, but there is a bug here: https://github.com/Qiskit/qiskit-terra/blob/80ca284048e2e69e1e9eb8b92112dccefeeb2f64/qiskit/visualization/text.py#L706

We omit array parameters from drawing only if all parameters are ndarrays. We can fix that here or in its own issue.

Reply:
Luciano actually just fixed that and pushed to this PR, so we're all good.